lasagne/layers/merge.py

```diff
@@ -312,9 +327,20 @@ def __init__(self, incomings, merge_function, cropping=None, **kwargs):
 
     def get_output_shape_for(self, input_shapes):
         input_shapes = autocrop_array_shapes(input_shapes, self.cropping)
-        if any(shape != input_shapes[0] for shape in input_shapes):
+        # Infer as much of the shape as possible, avoiding Nones
+        output_shape = tuple(next((s[i] for s in input_shapes
+                                   if s[i] is not None), None)
+                             for i in range(len(input_shapes[0])))
+
```

Comment:
This might be a little cleaner as 

        output_shape = [next((s for s in shape if s is not None), None)
                        for shape in zip(*input_shapes)]

Although it still seems pretty obtuse to me, a for loop might be better for readability.